### PR TITLE
New week is only determined through current day + map seed

### PR DESCRIPTION
**Pull request by vincent-grosbois**
This make new week deterministic once the map is created
It may not 100% be what we want as in heroes 2 I think it's also partly
random once the map is created

### CODE REVIEW

**Comment by Branikolog** [comment]
> This make new week deterministic once the map is created It may not 100% be what we want as in heroes 2 I think it's also partly random once the map is created

Hi, @vincent-grosbois !
In the OG if we load autosave - the week will always be the same. But if we load and replay the game from a more distant point (like 3+ days) the week changes... 
Not sure about H2, but comparing with H3 autosave can even bring other week in a case AI actions were completely different.

**Comment by vincent-grosbois** [comment]
> > This make new week deterministic once the map is created It may not 100% be what we want as in heroes 2 I think it's also partly random once the map is created
> 
> Hi, @vincent-grosbois ! In the OG if we load autosave - the week will always be the same. But if we load and replay the game from a more distant point (like 3+ days) the week changes... Not sure about H2, but comparing with H3 autosave can even bring other week in a case AI actions were completely different.

Indeed it's basically what I was  saying
now the question I had is _how_ this change of week exactly appears, or rather, if we can do something similar.
One thing I thought about (maybe) is to also hash based on the list of all heroes' stats

**Comment by ihhub** [comment]
Hi @vincent-grosbois , I think the original game depends on all neutral monsters on the map instead of heroes. Logically, spawning of neutrals relates on locations where other monsters stay. To prove my theory: run the original game, make a save, check the new week, load the save and hire extra hero or dismiss few, check the new week. I think it should remain the same.

**Comment by vincent-grosbois** [comment]
> Hi @vincent-grosbois , I think the original game depends on all neutral monsters on the map instead of heroes.

Ok so for the hashing part I can make it dependent also on the list / position / type of neutral monsters
How does it sound ?

**Comment by a1exsh** [comment]
> > This make new week deterministic once the map is created It may not 100% be what we want as in heroes 2 I think it's also partly random once the map is created
> 
> Hi, @vincent-grosbois ! In the OG if we load autosave - the week will always be the same. But if we load and replay the game from a more distant point (like 3+ days) the week changes... Not sure about H2, but comparing with H3 autosave can even bring other week in a case AI actions were completely different.

Do you mean replay these 3 days identically to the first try? In that case we should expect the same outcome, including the week announcement. Otherwise, it can easily happen that the random number generator is asked different number of times, thus affecting all further outcomes.

**Comment by Branikolog** [comment]
Hi, @a1exsh!

> Do you mean replay these 3 days identically to the first try? In that case we should expect the same outcome, including the week announcement. Otherwise, it can easily happen that the random number generator is asked different number of times, thus affecting all further outcomes.

Yes. In the OG I made a savepoint on tuesday, then simply skipped 3 days in a row and each reload was completely different week announcement.

**Comment by ihhub** [comment]
Hi @vincent-grosbois , yes. This is exactly what I'm thinking.

**Comment by ihhub** [comment]
> 
> 
> Hi, @a1exsh!
> 
> > Do you mean replay these 3 days identically to the first try? In that case we should expect the same outcome, including the week announcement. Otherwise, it can easily happen that the random number generator is asked different number of times, thus affecting all further outcomes.
> 
> Yes. In the OG I made a savepoint on tuesday, then simply skipped 3 days in a row and each reload was completely different week announcement.

It's because AI most likely fought neutrals. Make a map with no neutrals and try this.

**Comment by vincent-grosbois** [comment]
@ihhub you can review it then :)

File: src/fheroes2/world/world.h [hunk 1]
```diff
@@ -335,7 +335,8 @@ class World : protected fheroes2::Size
     std::vector<MapRegion> _regions;
     PlayerWorldPathfinder _pathfinder;
 
-    uint32_t _seed;
+    uint32_t _seed{ 0 }; // global seed for the map
+    size_t _week_seed; // global seed for the map, for this week
```
[the remark this answers is not in the document]
Comment (ihhub) [reply src/fheroes2/world/world.h:339]: Since we assigned the value for the above variable let's do the same for this member.

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @vincent-grosbois , I left few comments in this pull request. Could you please take a look?

**Comment by ihhub** [comment]
@vincent-grosbois , thank you so much for this useful improvement!

**Pull request merged**